### PR TITLE
[clangd] Let DefineOutline tweak handle member function templates

### PR DESCRIPTION
**Pull request by ckandeler**

### CODE REVIEW

**Comment by HighCommander4** [comment]
It looks like it was a deliberate design choice to disable this tweak for templates: https://reviews.llvm.org/D85310.

cc @kadircet, @hokein for any thoughts

**Comment by kadircet** [comment]
i remember mainly two concerns;
- we didn't want to surprise developers by breaking their builds as template definitions needs to be visible at instantiation sites. so if we keep the definition in the same file, just move it out-of-line, this should be fine.
- i remember AST was lacking information about template parameter names, maybe that's resolved now. This was causing issues as you need to print the whole outer template parameter list as-is, e.g. in the tests behavior looks like:
```cpp
template <typename U>
inline void Foo<T>::bar(const T& t, const U& u) {}
```
this won't compile as `T` isn't defined. you also need to print `template <typename T>` on top (and keep doing for rest of the outer decls).

if we can address both of these, i don't think there's any other reason to not have them

**Comment by ckandeler** [comment]
> ```c++
> template <typename U>
> inline void Foo<T>::bar(const T& t, const U& u) {}
> ```
> 
> this won't compile as `T` isn't defined. you also need to print `template <typename T>` on top (and keep doing for rest of the outer decls).

The patch does that already.
In the test, the outer template declaration is on the line above.

File: clang-tools-extra/clangd/refactor/tweaks/DefineOutline.cpp [hunk 1]
```diff
@@ -130,23 +129,40 @@ getFunctionSourceAfterReplacements(const FunctionDecl *FD,
...
-        if (!S.empty())
-          *S.rbegin() = '\n'; // Replace space with newline
-        TemplatePrefix.insert(0, S);
-      }
+      AddToTemplatePrefixIfApplicable(Parent, false);
     }
   }
 
+  AddToTemplatePrefixIfApplicable(FD, true);
   auto Source = QualifiedFunc->substr(FuncBegin, FuncEnd - FuncBegin + 1);
```
Comment (HighCommander4) [inline clang-tools-extra/clangd/refactor/tweaks/DefineOutline.cpp:163]: This is really a comment on the pre-existing code, but could you move this declaration of `Source` further up, to before `TemplatePrefix`?

File: clang-tools-extra/clangd/refactor/tweaks/DefineOutline.cpp [hunk 2]
```diff
@@ -130,23 +129,40 @@ getFunctionSourceAfterReplacements(const FunctionDecl *FD,
     return QualifiedFunc.takeError();
 
   std::string TemplatePrefix;
+  auto AddToTemplatePrefixIfApplicable = [&](const Decl *D, bool append) {
```
Comment (HighCommander4) [inline clang-tools-extra/clangd/refactor/tweaks/DefineOutline.cpp:132]: nit: UpperCamelCase for `append`

File: clang-tools-extra/clangd/refactor/tweaks/DefineOutline.cpp [hunk 3]
```diff
@@ -130,23 +129,40 @@ getFunctionSourceAfterReplacements(const FunctionDecl *FD,
...
-        Params->print(Stream, FD->getASTContext());
-        if (!S.empty())
-          *S.rbegin() = '\n'; // Replace space with newline
-        TemplatePrefix.insert(0, S);
-      }
+      AddToTemplatePrefixIfApplicable(Parent, false);
     }
   }
 
+  AddToTemplatePrefixIfApplicable(FD, true);
```
Comment (HighCommander4) [inline clang-tools-extra/clangd/refactor/tweaks/DefineOutline.cpp:162]: suggestion: remove the `append` parameter, always prepend, and call this before the loop over parents?

**Review by HighCommander4 — APPROVED** [review]
LGTM with minor comments addressed

File: clang-tools-extra/clangd/refactor/tweaks/DefineOutline.cpp [hunk 4]
```diff
@@ -443,13 +461,26 @@ class DefineOutline : public Tweak {
         SameFile = true;
 
         // Bail out if the template parameter is unnamed.
+        // FIXME: Is this really needed? It inhibits application on
```
[the remark this answers is not in the document]
Comment (ckandeler) [reply clang-tools-extra/clangd/refactor/tweaks/DefineOutline.cpp]: I've moved the comment down to the function template part.

**Review by HighCommander4 — APPROVED** [review]
Anyways, LGTM!

**Pull request merged**